pricegraph/src/graph/shortest_paths.rs

```diff
@@ -80,34 +133,63 @@ where
 
     /// Lists all nodes that can be reached from the source.
     pub fn connected_nodes(&self) -> Vec<G::NodeId> {
-        let mut node_indices: Vec<_> = self
-            .predecessor_store
-            .predecessors
-            .iter()
-            .enumerate()
-            .filter_map(|(i, &pre)| pre.map(|_| self.graph.from_index(i)))
-            .collect();
+        let mut node_indices: Vec<_> = match &self.predecessor_store {
+            PredecessorStore::Unbounded(predecessors, _) => {
+                nodes_from_predecessors(self.graph, &predecessors)
+            }
+            PredecessorStore::Bounded(predecessors_at_step, _) => {
+                let mut repeating_node_indices: Vec<_> = predecessors_at_step
+                    .iter()
+                    .map(|predecessors| nodes_from_predecessors(self.graph, &predecessors))
+                    .flatten()
+                    .collect();
+                repeating_node_indices.sort();
+                repeating_node_indices.dedup();
```

Comment:
Maybe insert into HashSet instead

Reply:
Why does it need to be sorted?

Reply:
In the current version of the code it's sorted so that `dedup` works correctly (see std docs on dedup).

Reply:
The return type requires a `Vec` so we would have to go from `impl Iterator -> HashSet -> Vec`. I think maybe we should:
- measure the performance impact of that change (since this would make reducing the orderbook even slower, which is currently the bottleneck).
- If the impact is significant, then maybe just group the two statements so it is clear why they are together:
```rust
{
    // NOTE: `dedup` requires a sorted vector.
   repeating_node_indices.sort();
   repeating_node_indices.dedup();
}
```

Reply:
Performance doesn't seem to change too much when using `HashSet`, worst performance decreases are in `Pricegraph::order_for_limit_price`, of about 4%.

Tested changes are some extra `Hash` trait bounds and:
```suggestion
                let mut unique_nodes = HashSet::new();
                for node in repeating_node_indices.into_iter() {
                   unique_nodes.insert(node)
                }
                unique_nodes.drain().collect()
```


<details>
<summary>Full results</summary>

```
Pricegraph::read        time:   [5.0849 ms 5.0906 ms 5.0985 ms]                              
                        change: [+1.8985% +2.0957% +2.3268%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 11 outliers among 500 measurements (2.20%)
  8 (1.60%) high mild
  3 (0.60%) high severe

Pricegraph::transitive_orderbook/5298183                                                                             
                        time:   [4.7446 ms 4.7504 ms 4.7564 ms]
                        change: [-5.2606% -4.9262% -4.6681%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 500 measurements (0.20%)
  1 (0.20%) high mild

Benchmarking Pricegraph::estimate_limit_price/100000000000000000: Warming up for 3.0000 s
Warning: Unable to complete 500 samples in 5.0s. You may wish to increase target time to 7.5s, enable flat sampling, or reduce sample count to 280.
Pricegraph::estimate_limit_price/100000000000000000                                                                             
                        time:   [57.816 us 57.912 us 58.011 us]
                        change: [+1.3288% +1.8919% +2.4802%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 500 measurements (1.60%)
  5 (1.00%) high mild
  3 (0.60%) high severe
Benchmarking Pricegraph::estimate_limit_price/1000000000000000000: Warming up for 3.0000 s
Warning: Unable to complete 500 samples in 5.0s. You may wish to increase target time to 7.9s, enable flat sampling, or reduce sample count to 280.
Pricegraph::estimate_limit_price/1000000000000000000                                                                             
                        time:   [60.097 us 60.179 us 60.265 us]
                        change: [-1.3895% -0.8448% -0.3410%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 14 outliers among 500 measurements (2.80%)
  9 (1.80%) high mild
  5 (1.00%) high severe
Pricegraph::estimate_limit_price/10000000000000000000                                                                            
                        time:   [106.58 us 106.68 us 106.77 us]
                        change: [+1.5183% +1.8297% +2.0602%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 142 outliers among 500 measurements (28.40%)
  106 (21.20%) low severe
  12 (2.40%) low mild
  10 (2.00%) high mild
  14 (2.80%) high severe
Pricegraph::estimate_limit_price/100000000000000000000                                                                            
                        time:   [200.96 us 201.21 us 201.48 us]
                        change: [+0.4806% +0.7399% +0.9579%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 500 measurements (1.40%)
  6 (1.20%) high mild
  1 (0.20%) high severe
Pricegraph::estimate_limit_price/1000000000000000000000                                                                            
                        time:   [453.78 us 454.39 us 455.03 us]
                        change: [+0.7609% +1.0751% +1.3240%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 18 outliers among 500 measurements (3.60%)
  11 (2.20%) high mild
  7 (1.40%) high severe

Benchmarking Pricegraph::order_for_limit_price/200: Warming up for 3.0000 s
Warning: Unable to complete 500 samples in 5.0s. You may wish to increase target time to 6.6s, enable flat sampling, or reduce sample count to 300.
Pricegraph::order_for_limit_price/200                                                                             
                        time:   [50.740 us 50.821 us 50.903 us]
                        change: [-1.4101% -0.7140% -0.1848%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 500 measurements (1.40%)
  3 (0.60%) high mild
  4 (0.80%) high severe
Pricegraph::order_for_limit_price/190                                                                            
                        time:   [136.99 us 137.15 us 137.32 us]
                        change: [+3.4241% +3.7101% +3.9312%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 136 outliers among 500 measurements (27.20%)
  89 (17.80%) low severe
  21 (4.20%) low mild
  9 (1.80%) high mild
  17 (3.40%) high severe
Pricegraph::order_for_limit_price/180                                                                            
                        time:   [220.40 us 220.64 us 220.89 us]
                        change: [+4.4884% +4.6582% +4.8224%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 6 outliers among 500 measurements (1.20%)
  4 (0.80%) high mild
  2 (0.40%) high severe
Pricegraph::order_for_limit_price/150                                                                            
                        time:   [452.49 us 453.31 us 454.37 us]
                        change: [+2.9383% +3.2534% +3.5734%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 500 measurements (0.60%)
  1 (0.20%) high mild
  2 (0.40%) high severe
Pricegraph::order_for_limit_price/100                                                                            
                        time:   [489.42 us 490.06 us 490.73 us]
                        change: [+1.9823% +2.1616% +2.3452%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 500 measurements (0.40%)
  2 (0.40%) high severe
```

</details>

Reply:
What if you use `into_iter` instead of `drain`?

Reply:
It doesn't seem to make much of a difference, here is the bench against the current PR for comparison.

<details>
<summary>Full results</summary>

```
Benchmarking Pricegraph::read: Warming up for 3.0000 s
Warning: Unable to complete 50 samples in 5.0s. You may wish to increase target time to 6.4s, enable flat sampling, or reduce sample count to 30.
Pricegraph::read        time:   [5.0343 ms 5.0500 ms 5.0644 ms]                             
                        change: [-3.2447% -1.9703% -0.5280%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 10 outliers among 50 measurements (20.00%)
  7 (14.00%) low mild
  2 (4.00%) high mild
  1 (2.00%) high severe

Benchmarking Pricegraph::transitive_orderbook/5298183: Warming up for 3.0000 s
Warning: Unable to complete 50 samples in 5.0s. You may wish to increase target time to 6.4s, enable flat sampling, or reduce sample count to 30.
Pricegraph::transitive_orderbook/5298183                                                                            
                        time:   [4.8342 ms 4.8490 ms 4.8639 ms]
                        change: [-4.1812% -2.8137% -1.6666%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 50 measurements (6.00%)
  2 (4.00%) high mild
  1 (2.00%) high severe

Pricegraph::estimate_limit_price/100000000000000000                                                                           
                        time:   [57.368 us 57.601 us 57.796 us]
                        change: [-7.8755% -5.8115% -4.5116%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 50 measurements (12.00%)
  5 (10.00%) low mild
  1 (2.00%) high mild
Pricegraph::estimate_limit_price/1000000000000000000                                                                           
                        time:   [61.133 us 61.373 us 61.615 us]
                        change: [-5.0646% -4.3206% -3.5991%] (p = 0.00 < 0.05)
                        Performance has improved.
Pricegraph::estimate_limit_price/10000000000000000000                                                                           
                        time:   [107.26 us 108.37 us 109.60 us]
                        change: [-1.0815% -0.6203% -0.0586%] (p = 0.02 < 0.05)
                        Change within noise threshold.
Found 2 outliers among 50 measurements (4.00%)
  2 (4.00%) high severe
Pricegraph::estimate_limit_price/100000000000000000000                                                                           
                        time:   [203.57 us 204.02 us 204.44 us]
                        change: [+0.0362% +0.3729% +0.7059%] (p = 0.03 < 0.05)
                        Change within noise threshold.
Found 8 outliers among 50 measurements (16.00%)
  1 (2.00%) low severe
  6 (12.00%) low mild
  1 (2.00%) high severe
Pricegraph::estimate_limit_price/1000000000000000000000                                                                           
                        time:   [455.62 us 456.77 us 457.85 us]
                        change: [+1.2362% +1.6203% +2.0236%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 50 measurements (14.00%)
  5 (10.00%) low mild
  1 (2.00%) high mild
  1 (2.00%) high severe

Pricegraph::order_for_limit_price/200                                                                           
                        time:   [51.123 us 51.361 us 51.641 us]
                        change: [-3.4017% -2.7420% -2.0780%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 50 measurements (24.00%)
  1 (2.00%) high mild
  11 (22.00%) high severe
Pricegraph::order_for_limit_price/190                                                                           
                        time:   [132.39 us 132.64 us 132.95 us]
                        change: [-0.4205% -0.2325% -0.0333%] (p = 0.03 < 0.05)
                        Change within noise threshold.
Found 8 outliers among 50 measurements (16.00%)
  2 (4.00%) low severe
  1 (2.00%) high mild
  5 (10.00%) high severe
Pricegraph::order_for_limit_price/180                                                                           
                        time:   [217.06 us 217.84 us 218.65 us]
                        change: [+0.0761% +0.4669% +0.8369%] (p = 0.02 < 0.05)
                        Change within noise threshold.
Found 10 outliers among 50 measurements (20.00%)
  3 (6.00%) low severe
  5 (10.00%) low mild
  2 (4.00%) high severe
Pricegraph::order_for_limit_price/150                                                                           
                        time:   [449.34 us 450.96 us 452.58 us]
                        change: [+1.5642% +2.2870% +3.1362%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 50 measurements (4.00%)
  2 (4.00%) high severe
Pricegraph::order_for_limit_price/100                                                                           
                        time:   [499.42 us 511.71 us 525.95 us]
                        change: [+3.0038% +4.3750% +5.8430%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 18 outliers among 50 measurements (36.00%)
  2 (4.00%) low severe
  5 (10.00%) low mild
  1 (2.00%) high mild
  10 (20.00%) high sever	
```

</details>